tests/test_dataclasses.py

```diff
@@ -633,3 +633,93 @@ class MyDataclass:
+
+
+def test_override_builtin_dataclass():
+    @dataclasses.dataclass
+    class File:
+        hash: str
+        name: Optional[str]
+        size: int
+        content: Optional[bytes] = None
+
+    FileChecked = pydantic.dataclasses.dataclass(File)
+    f = FileChecked(hash='xxx', name=b'whatever.txt', size='456')
+    assert f.name == 'whatever.txt'
+    assert f.size == 456
+
+    with pytest.raises(ValidationError) as e:
+        FileChecked(hash=[1], name='name', size=3)
+    assert e.value.errors() == [{'loc': ('hash',), 'msg': 'str type expected', 'type': 'type_error.str'}]
+
+
+def test_override_builtin_dataclass_2():
+    @dataclasses.dataclass
+    class Meta:
+        modified_date: Optional[datetime]
+        seen_count: int
+
+    @pydantic.dataclasses.dataclass
+    @dataclasses.dataclass
+    class File(Meta):
+        filename: str
+
+    f = File(filename=b'thefilename', modified_date='2020-01-01T00:00', seen_count='7')
+    assert f.filename == 'thefilename'
+    assert f.modified_date == datetime(2020, 1, 1, 0, 0)
+    assert f.seen_count == 7
+
+
+def test_override_builtin_dataclass_nested():
+    @dataclasses.dataclass
+    class Meta:
+        modified_date: Optional[datetime]
+        seen_count: int
+
+    @dataclasses.dataclass
+    class File:
+        filename: str
+        meta: Meta
+
+    FileChecked = pydantic.dataclasses.dataclass(File)
+    f = FileChecked(filename=b'thefilename', meta=Meta(modified_date='2020-01-01T00:00', seen_count='7'))
+    assert f.filename == 'thefilename'
+    assert f.meta.modified_date == datetime(2020, 1, 1, 0, 0)
+    assert f.meta.seen_count == 7
+
+    with pytest.raises(ValidationError) as e:
+        FileChecked(filename=b'thefilename', meta=Meta(modified_date='2020-01-01T00:00', seen_count=['7']))
+    assert e.value.errors() == [
+        {'loc': ('meta', 'seen_count'), 'msg': 'value is not a valid integer', 'type': 'type_error.integer'}
+    ]
+
+
+def test_override_builtin_dataclass_nested_schema():
+    @dataclasses.dataclass
+    class Meta:
+        modified_date: Optional[datetime]
+        seen_count: int
+
+    @dataclasses.dataclass
+    class File:
+        filename: str
+        meta: Meta
+
+    FileChecked = pydantic.dataclasses.dataclass(File)
+    assert FileChecked.__pydantic_model__.schema() == {
+        'definitions': {
+            'Meta': {
+                'properties': {
+                    'modified_date': {'format': 'date-time', 'title': 'Modified ' 'Date', 'type': 'string'},
+                    'seen_count': {'title': 'Seen Count', 'type': 'integer'},
+                },
+                'required': ['modified_date', 'seen_count'],
+                'title': 'Meta',
+                'type': 'object',
+            }
+        },
+        'properties': {'filename': {'title': 'Filename', 'type': 'string'}, 'meta': {'$ref': '#/definitions/Meta'}},
+        'required': ['filename', 'meta'],
+        'title': 'File',
+        'type': 'object',
+    }
```

Comment:
can we add a case like:

```py
@dataclasses.dataclass
class File:
    filename: str
    meta: Meta


class Foo(BaseModel):
    file: File

...
```

I know this should work, but this is what I often want to do so best to be explicit that it works.